lib/cabal-project-parser.nix

```diff
@@ -95,10 +95,90 @@ let
           otherText = "\nsource-repository-package\n" + block;
         }
         else {
-          sourceRepo = extractRepoData cabalProjectFileName lookupSha256 attrs;
+          sourceRepo = extractSourceRepoPackageData cabalProjectFileName lookupSha256 attrs;
           otherText = pkgs.lib.strings.concatStringsSep "\n" x.snd;
         };
 
+  parseSourceRepositoryPackages = cabalProjectFileName: lookupSha256: source-repo-override: projectFile:
+    let
+      blocks = pkgs.lib.splitString "\nsource-repository-package\n" ("\n" + projectFile);
+      initialText = pkgs.lib.lists.take 1 blocks;
+      repoBlocks = builtins.map (parseBlock cabalProjectFileName lookupSha256) (pkgs.lib.lists.drop 1 blocks);
+      overrideSourceRepo = sourceRepo: (source-repo-override.${sourceRepo.url} or (pkgs.lib.id)) sourceRepo;
+    in {
+      sourceRepos = pkgs.lib.lists.map (block: overrideSourceRepo block.sourceRepo) repoBlocks;
+      otherText = pkgs.lib.strings.concatStringsSep "\n" (
+        initialText
+        ++ (builtins.map (x: x.otherText) repoBlocks));
+    };
+
+  # Parse a repository
+  parseRepositoryBlock = cabalProjectFileName: sha256map: cabal-install: nix-tools: block:
+    let
+      lines = pkgs.lib.splitString "\n" block;
+      x = span (pkgs.lib.strings.hasPrefix " ") (__tail lines);
+      attrs = parseBlockLines x.fst;
+    in rec {
+      name = __head lines;
+      repo = attrs;
+      home =
+        pkgs.evalPackages.runCommandLocal name {
+          nativeBuildInputs = [ cabal-install pkgs.evalPackages.curl nix-tools ];
+          LOCALE_ARCHIVE = pkgs.lib.optionalString (pkgs.evalPackages.stdenv.buildPlatform.libc == "glibc") "${pkgs.evalPackages.glibcLocales}/lib/locale/locale-archive";
+          LANG = "en_US.UTF-8";
+          outputHashMode = "recursive";
+          outputHashAlgo = "sha256";
+          outputHash = attrs."--sha256" or sha256map.${attrs.url};
+        } ''
+            mkdir -p $out/.cabal
+            cat <<EOF > $out/.cabal/config
```

Comment:
Should we be including a `index-state` pin if there is one in the `cabal.project`? That would let this be somewhat deterministic even for repositories with non-fixed content.

Reply:
If that would work, it would be interesting to see a test with, say, `head.hackage`!

Reply:
It seems so far we have two types of repositories:

`hackage-server` repos where I think the `index-state` will be needed and we should probably add support for per repository `index-state`.  I think this should be done in another PR when we have a requirement though.

`hackage-overlay-repo-tool` based repos running of some GitHub or gitlab source.  In these it is likely the timestamps in the index are all the same (all the files in https://ghc.gitlab.haskell.org/head.hackage/01-index.tar.gz right now are dated `11 Feb 17:34`), so it is more useful to be able to specify the git commit in the `url`.  I'm not sure If this is possible with `head.hackage`, but it is with `hackage-ghcjs-overlay`.

The good news is that this seems to work:

```
repository ghcjs-overlay
  url: https://raw.githubusercontent.com/input-output-hk/hackage-overlay-ghcjs/bfc363b9f879c360e0a0460ec0c18ec87222ec32
  secure: True
  root-keys:
  key-threshold: 0
  --sha256: sha256-g9xGgJqYmiczjxjQ5JOiK5KUUps+9+nlNGI/0SpSOpg=

repository head.hackage.ghc.haskell.org
   url: https://ghc.gitlab.haskell.org/head.hackage/
   secure: True
   key-threshold: 3
   root-keys:
       f76d08be13e9a61a377a85e2fb63f4c5435d40f8feb3e12eb05905edb8cdea89
       26021a13b401500c8eb2761ca95c61f2d625bfef951b939a8124ed12ecf07329
       7541f32a4ccca4f97aea3b22f5e593ba2c0267546016b992dfadcd2fe944e55d
   --sha256: sha256-IZhP4UGOPv/E89xP89h7RLcddbOV5jpAu9sDVrpfp90=
```

But without a git commit in the url the `head.hackage` one will break at some point.  The commit hash in the `ghcjs-overlay` `url` is from the `gh-pages` branch, perhaps there is an equivalent for gitlab.